### PR TITLE
scripts/release-notes: report missing RNs, don't skip standalone commits

### PR DESCRIPTION
**Pull request by knz**
(All commits but the last from #42344 and prior)

Prior to this patch, the release note script would skip over
standalone commits (without a PR) when extracting release notes.
This was defective, as the git history does have a few example
standalone commits, with release note but without PR.

Additionally, the script now reports PRs that are missing release
notes *or only use 'Release note: none'*, which may indicate
that the engineer did not bother filling in a note despite the
presence of a user-facing change.

For example:

```
$ python3 scripts/release-notes.py \
    --from v19.2.0-beta.20190930 \
    --until provisional_201911080435_v19.2.0-rc.5
```

Outputs (snippet):

- Andrei Matei:
  - 2019-10-04 [#41303][#41303] [ca32e6ff0][ca32e6ff0] sql: remove dead code [NO RELEASE NOTE]
  - 2019-10-05 [#41307][#41307] [1c99165c3][1c99165c3] sql: fix various problematic uses of the txn in DistSQL flows (4 commits)
  - 2019-10-28 [#41935][#41935] [bad8e55bb][bad8e55bb] settings/cluster: introducing the 19.2 cluster version [NO RELEASE NOTE]

In this example, the first PR has no release note but does not deserve
one. The second has a release note. The third is marked with "Release
note: None" but **there should really have been a release note**: the
introduction of the major cluster version is an important user-facing
change and should be announced.

By highlighting the missing release notes in this way, the script
gives an opportunity to the doc writer(s) to find additional
user-facing changes that need to be documented.

[#41303]: https://github.com/cockroachdb/cockroach/pull/41303
[#41307]: https://github.com/cockroachdb/cockroach/pull/41307
[#41935]: https://github.com/cockroachdb/cockroach/pull/41935
[ca32e6ff0]: https://github.com/cockroachdb/cockroach/commit/ca32e6ff0
[1c99165c3]: https://github.com/cockroachdb/cockroach/commit/1c99165c3
[bad8e55bb]: https://github.com/cockroachdb/cockroach/commit/bad8e55bb

### CODE REVIEW

**Review by jseldess — COMMENTED** [review]
Thank you for taking this on, @knz!  The code changes LGTM though I'd suggest asking another dev to review as well. But otherwise, I have a few questions. My understanding is that these changes will capture:

1. Commits with release notes texts but without any parent PR. 
2. PRs for which all commits have "Release note: none" or similar.

For 1, I definitely agree we should capture these. But can you help me understand how it's possible for a commit to not be associated with a PR?

For 2, I'm worried that we'll be putting too much weight on the docs writers who generate our release notes to verify whether or not these cases are valid. It will be very time-consuming to analyze each such PR, and our codebase is so broad that no single writer will be able to even judge all cases without tracking down authors, etc. Another approach we could take is to stop defaulting to `Release note: none` and start requiring engineers to explicitly write release note texts. This will add overhead for engs, but it will also mean that we can trust "Release note: none", at least in more cases. Thoughts? 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42345#-:-LtL6VH88Zw0zlPAQQxD:b-by7ar8)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jseldess)

**Comment by knz** [comment]
> For 1, I definitely agree we should capture these. But can you help me understand how it's possible for a commit to not be associated with a PR?

If we temporarily disable the checkbox "force a merge commit", then the green button can slide the git history to capture a PR's commit without registering the PR number in the history. So there would likely be a PR but we can't find it in the script.

It has happened for just 2 commits since summer 2018, so it's rather infrequent.

>  For 2, [...] 

So first just to clarify, what the script does is add a marker "[NO RELEASE NOTE]" to the relevant **per-contributor changes**, which is the section that *you have recommended your team to hide by default* with `--hide-per-contributor-section`.

So this is not costing your team anything (yet) until we decide what to do with it.

> It will be very time-consuming to analyze each such PR, and our codebase is so broad that no single writer will be able to even judge all cases without tracking down authors, etc. 

Agreed. I think these things should be reviewed perhaps out-of-band from the release weeks, perhaps by the TLs of each team.

> Another approach we could take is to stop defaulting to Release note: none and start requiring engineers to explicitly write release note texts. 

But there are cases where there is indeed no user-facing change. What to write in this case?

My opinion on this is that I am changing the tool in order to *create an opportunity* to solve a problem that exists and for which we don't have a solution. I am not yet 100% sure how to actually use the tool to solve the problem.

I would really like if the TLs would own the problem for their teams, which would be scalable over time. But I don't know how to make this happen yet.

**Review by jseldess — APPROVED** [review]
Thanks for the clarifications, @knz. Definitely LGTM then. If you'd like me to test this out before you merge, I'll have time after the release. Your call.

> But there are cases where there is indeed no user-facing change. What to write in this case?

Yes, I meant that in those cases the dev would write `Release note: none`. The change is that `Release note: none` would no longer be the default. It would have to be explicitly added. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/42345#-:-LtLCl6P1DnSa3g3yUzj:b2iv2qu)* status: :shipit: complete! 0 of 0 LGTMs obtained

**Comment by knz** [comment]
> Yes, I meant that in those cases the dev would write Release note: none. The change is that Release note: none would no longer be the default. It would have to be explicitly added.

Oh we made that change already a couple weeks ago.

**Comment by jseldess** [comment]
Nice. I didn't know that.

**Comment by knz** [comment]
@chrisseto if you could review the python code change, that would be useful.

**Comment by knz** [comment]
We don't have standalone commits any more so we don't need this.

**Pull request closed without merging**
